gkr/src/prover/gkr.rs

```diff
@@ -1,58 +1,36 @@
 //! This module implements the core GKR IOP.
 
 use circuit::Circuit;
-use gkr_field_config::GKRFieldConfig;
-use mpi_config::MPIConfig;
-use polynomials::MultiLinearPolyExpander;
+use gkr_engine::{ExpanderDualVarChallenge, FieldEngine, MPIConfig, MPIEngine, Transcript};
 use sumcheck::{sumcheck_prove_gkr_layer, ProverScratchPad};
-use transcript::Transcript;
 use utils::timer::Timer;
 
 // FIXME
 #[allow(clippy::type_complexity)]
-pub fn gkr_prove<C: GKRFieldConfig, T: Transcript<C::ChallengeField>>(
-    circuit: &Circuit<C>,
-    sp: &mut ProverScratchPad<C>,
-    transcript: &mut T,
+pub fn gkr_prove<F: FieldEngine>(
+    circuit: &Circuit<F>,
+    sp: &mut ProverScratchPad<F>,
+    transcript: &mut impl Transcript<F::ChallengeField>,
     mpi_config: &MPIConfig,
-) -> (
-    C::ChallengeField,
-    Vec<C::ChallengeField>,
-    Option<Vec<C::ChallengeField>>,
-    Vec<C::ChallengeField>,
-    Vec<C::ChallengeField>,
-) {
+) -> (F::ChallengeField, ExpanderDualVarChallenge<F>) {
     let layer_num = circuit.layers.len();
 
-    let mut rz0 = vec![];
-    let mut rz1 = None;
-    let mut r_simd = vec![];
-    let mut r_mpi = vec![];
-    for _ in 0..circuit.layers.last().unwrap().output_var_num {
-        rz0.push(transcript.generate_challenge_field_element());
-    }
-
-    for _ in 0..C::get_field_pack_size().trailing_zeros() {
-        r_simd.push(transcript.generate_challenge_field_element());
-    }
-
-    for _ in 0..mpi_config.world_size().trailing_zeros() {
-        r_mpi.push(transcript.generate_challenge_field_element());
-    }
+    let mut challenge = ExpanderDualVarChallenge::sample_from_transcript(
```

Comment:
was wondering why we want to sample challenge directly from FS RO, now it seems reasonable to start proving the output layer